[pytorch] Migrating index_add cuda to ATen

ggoossen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30573 [pytorch] Migrating index_add cuda to ATen**

Mostly just moved code.
Index dim and number of indices checks are added to make checks idential to index_add_cpu_

Differential Revision: [D18749922](https://our.internmc.facebook.com/intern/diff/D18749922/)

ggoossen committed:
[pytorch] Migrating index_add cuda to ATen

Mostly just moved code.
Index dim and number of indices checks are added to make checks idential to index_add_cpu_

Differential Revision: [D18749922](https://our.internmc.facebook.com/intern/diff/D18749922/)

[ghstack-poisoned]

ggoossen committed:
Update on "[pytorch] Migrating index_add cuda to ATen"


Mostly just moved code.
Index dim and number of indices checks are added to make checks idential to index_add_cpu_

Differential Revision: [D18749922](https://our.internmc.facebook.com/intern/diff/D18749922/)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit f6c25f75:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=f6c25f75e168c75dc6fa9ecc41111d4419b54697).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4236544) pytorch_xla_linux_xenial_py3_6_clang7_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=85738677) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=85738677))
<details>
<summary>
<code>Jan 14 11:46:46 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff</code>
</summary>

```
Jan 14 11:46:46 + XDIR=/var/lib/jenkins/workspace/xla/scripts/.. 
Jan 14 11:46:46 + PTDIR=/var/lib/jenkins/workspace/xla/scripts/../.. 
Jan 14 11:46:46 + python /var/lib/jenkins/workspace/xla/scripts/cond_patch.py /var/lib/jenkins/workspace/xla/scripts/../torch_patches /var/lib/jenkins/workspace/xla/scripts/../.. 
Jan 14 11:46:46 1 out of 1 hunk FAILED 
Jan 14 11:46:46 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Jan 14 11:46:46 Traceback (most recent call last): 
Jan 14 11:46:46   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 67, in <module> 
Jan 14 11:46:46     patch_repo(args) 
Jan 14 11:46:46   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 49, in patch_repo 
Jan 14 11:46:46     raise RuntimeError('Failed to apply patch: {}'.format(ppath)) 
Jan 14 11:46:46 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Jan 14 11:46:46 + cleanup 
Jan 14 11:46:46 + retcode=1 
Jan 14 11:46:46 + set +x 
Jan 14 11:46:46 =================== sccache compilation log =================== 
Jan 14 11:46:46 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 14 11:46:46 Compile requests                 1 
Jan 14 11:46:46 Compile requests executed        0 
Jan 14 11:46:46 Cache hits                       0 
Jan 14 11:46:46 Cache misses                     0 
Jan 14 11:46:46 Cache timeouts                   0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

ggoossen committed:
Update on "[pytorch] Migrating index_add cuda to ATen"


Mostly just moved code.
Index dim and number of indices checks are added to make checks idential to index_add_cpu_

Differential Revision: [D18749922](https://our.internmc.facebook.com/intern/diff/D18749922/)

[ghstack-poisoned]

ggoossen commented:
Test failures appear unrelated.

Did a quick benchmark,
```
print(timeit.timeit(
    "dst.index_add_(index=index, dim=0, source=value)",
    setup="""
import torch
dst = torch.rand((1024, 32), device='cuda')
index = torch.randint(1024, (16*1024, ), device='cuda')
value = torch.rand((16*1024, 32), device='cuda')
    """,
    number=1000,
))

print(timeit.timeit(
    "dst.index_add_(index=index, dim=0, source=value)",
    setup="""
import torch
dst = torch.rand((1024*1024, 32), device='cuda')
index = torch.randint(1024*1024, (16*1024*1024, ), device='cuda')
value = torch.rand((16*1024*1024, 32), device='cuda')
    """,
    number=1000,
))

print(timeit.timeit(
    "dst.index_add_(index=index, dim=0, source=value)",
    setup="""
import torch
dst = torch.rand((1024*1024*32), device='cuda')
index = torch.randint(1024*1024*32, (16*1024*1024*32, ), device='cuda')
value = torch.rand((16*1024*1024*32), device='cuda')
    """,
    number=1000,
))
```
Results are the about the same:
Before:
> 0.017345599830150604                                                                         
> 0.018049221485853195     
> 16.25419460888952

After:
> 0.016109352931380272
> 0.01672299951314926
> 16.260623913258314

The 1000x difference between second and third result is odd, and would suggest the third case is far from optimal. But obviously unrelated to this change.

gchanan commented:
@VitalyFedyunin can you take a look?

VitalyFedyunin commented:
Sure!

ggoossen commented:
ping @VitalyFedyunin

VitalyFedyunin commented on aten/src/ATen/native/cuda/Indexing.cu:
```diff
@@ -245,7 +247,325 @@ void index_put_accum_kernel(Tensor & self, TensorList indices, const Tensor & va
+
+  if (mismatch) {
+    static bool warningShown = false;
+    if (!warningShown) {
```
Please use `TORCH_WARN_ONCE` instead of this construction.

ggoossen replied:
fixed.

VitalyFedyunin commented on aten/src/ATen/native/cuda/Indexing.cu:
```diff
@@ -245,7 +247,325 @@ void index_put_accum_kernel(Tensor & self, TensorList indices, const Tensor & va
+    return self;
+  }
+  const cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  int indContig = index.is_contiguous();
```
bool?

ggoossen replied:
changed to bool.

VitalyFedyunin reviewed (changes requested):
Looks good, just requires couple tiny changes.

ggoossen committed:
Update on "[pytorch] Migrating index_add cuda to ATen"


Mostly just moved code.
Index dim and number of indices checks are added to make checks idential to index_add_cpu_

Differential Revision: [D18749922](https://our.internmc.facebook.com/intern/diff/D18749922/)

[ghstack-poisoned]

ggoossen committed:
Update on "[pytorch] Migrating index_add cuda to ATen"


Mostly just moved code.
Index dim and number of indices checks are added to make checks idential to index_add_cpu_

Differential Revision: [D18749922](https://our.internmc.facebook.com/intern/diff/D18749922/)

[ghstack-poisoned]

zhangguanheng66 commented:
I have submitted request to revert this PR because it breaks several tests on the master branch. I also see the failures on diff. We have to investigate the reasons behind this.
